### PR TITLE
Keep small bigints small

### PR DESCRIPTION
**Pull request by erights**
This PR only changes the encoding of positive bigints so that positive small bigints have small encodings.

I'm not planning to do the same for small negative bigints, but it would be nice if a similar trick is possible.

### CODE REVIEW

**Review by FUDCo — CHANGES_REQUESTED** [review]
You should update the comment that describes the bigint encoding to reflect this change.

**Comment by mhofman** [comment]
After considering alternative encodings, this is a fairly flexible one. It took me a minutes to realize it shortens virtually all non extremely large bigints. I second a comment explaining this.

For future memory. Other encodings are possible using control chars right after the `p` (or `n`) that sort in the right lexical order.

**Comment by gibson042** [comment]
It looks like this is approximating a decimal [Elias delta coding](https://en.wikipedia.org/wiki/Elias_delta_coding), and probably should just go all the way to that. I'll try to provide an example later.

**Comment by erights** [comment]
> Elias delta coding

Is it sort-order preserving? The hack they recommend for negatives is not, without mentioning it, so it seems that is not a driving goal.

In any case, our remaining problem is a sort-order preserving encoding on the negative bitints, where small negative bigints have small encodings.

**Comment by erights** [comment]
> You should update the comment that describes the bigint encoding to reflect this change.

Done. PTAL.

File: packages/store/src/patterns/encodeKey.js
Comment (mhofman) [inline packages/store/src/patterns/encodeKey.js]: Makes me realize that we allow longer negative integers than positive ones. `lenlen` max value is `9` where we have a padding of the length to 10 digits for negative numbers. Since `0` is unused for `lenlen`, maybe a quick fix is to encode `${lenlen-1}` ?

Comment (mhofman) [reply packages/store/src/patterns/encodeKey.js]: Also means the negative digit `length` description is wrong, and you can't produce negative bingints with `11 ** 10` or more digits.

Comment (erights) [reply packages/store/src/patterns/encodeKey.js]: Done!

Comment (erights) [reply packages/store/src/patterns/encodeKey.js]: > Also means the negative digit length description is wrong, and you can't produce negative bingints with 11 ** 10 or more digits.

Could you suggest a correct description? Thanks.

File: packages/store/src/patterns/encodeKey.js
Comment (mhofman) [inline packages/store/src/patterns/encodeKey.js]: I maintain that it's smaller than `11 ** 10 ** 10` ;)

Comment (erights) [reply packages/store/src/patterns/encodeKey.js]: `10 ** 10 ** 11` ;)

Done.

**Review by FUDCo — APPROVED** [review]
It's swell!

**Comment by gibson042** [comment]
> > Elias delta coding
> 
> Is it sort-order preserving? The hack they recommend for negatives is not, without mentioning it, so it seems that is not a driving goal.
> 
> In any case, our remaining problem is a sort-order preserving encoding on the negative bitints, where small negative bigints have small encodings.

Yes, it can be. Here's one possibility that eliminates about three characters from the representation of each positive integer and is symmetrically sized for negative integers: https://jsfiddle.net/o6vw2nfd/ (run with the console open for visibility)
<details open><summary><code>bigintToDBEntryKey</code> implementation</summary>

```js
const bigintToDBEntryKey = n => {
  const nn = n < 0n ? -n : n;
  const nDigits = nn.toString().length;
  const lDigits = nDigits.toString().length;
  if (n < 0n) {
    return `n${

      // A zero for each digit beyond the first
      // in the decimal *count* of decimal digits.
      "0".repeat(lDigits - 1)}${

      // The count of digits, offset to be a
      // number of lDigits digits
      // that never starts with zero.
      // * lDigits=1: nDigits 1-9 => 9-1
      // * lDigits=2: nDigits 10-99 => 99-10
      // * lDigits=3: nDigits 100-999 => 999-100
      // * lDigits=4: nDigits 1000-9999 => 9999-1000
      // * ...
      10**lDigits + 10**(lDigits-1) - 1 - nDigits}${

      // The digits in a complementary representation
      // for reverse sorting.
      (10n**BigInt(nDigits) - 1n + n).toString().padStart(nDigits, "0")
      /*or e.g. 10n**nDigits + 10n**(nDigits-1n) - 1n + n*/}`;
  } else if (n === 0n) {
    return `o`;
  } else {
    return `p${

      // A nine for each digit beyond the first
      // in the decimal *count* of decimal digits.
      "9".repeat(lDigits - 1)}${

      // The count of digits,
      // offset to never start with a nine
      // and padded to lDigits digits.
      // * lDigits=1: nDigits 1-9 => 0-8
      // * lDigits=2: nDigits 10-99 => 00-89
      // * lDigits=3: nDigits 100-999 => 000-899
      // * lDigits=4: nDigits 1000-9999 => 0000-8999
      // * ...
      (nDigits - 10**(lDigits - 1)).toString().padStart(lDigits, "0")}${

      // The digits.
      n}`;
  }
};
```

<details><summary>deserialization</summary>

```js
const dbEntryKeyToBigint = k => {
  let ch = k[0];
  let rem = k.slice(1);
  switch (ch) {
    case "o": {
      if (rem.length > 0) throw new Error(`"o" must stand alone: ${k}`);
      return 0n;
    }
    case "n": {
      const lDigits = rem.match(/^0*/)[0].length + 1;
      rem = rem.slice(lDigits - 1);
      if (rem.length < lDigits) throw new Error(`incomplete digit count: ${k}`);
      const snDigits = rem.slice(0, lDigits);
      rem = rem.slice(lDigits);
      if (!/^[0-9]*$/.test(snDigits)) throw new Error(`invalid nDigits: ${k}`);
      const cnDigits = parseInt(snDigits, 10);
      const nDigits = 10**lDigits + 10**(lDigits-1) - 1 - cnDigits;
      if (rem.length !== nDigits) throw new Error(`digit count mismatch: ${k}`);
      const cn = BigInt(rem);
      const n = cn - 10n**BigInt(nDigits) + 1n;
      return n;
    }
    case "p": {
      const lDigits = rem.match(/^9*/)[0].length + 1;
      rem = rem.slice(lDigits - 1);
      if (rem.length < lDigits) throw new Error(`incomplete digit count: ${k}`);
      const snDigits = rem.slice(0, lDigits);
      rem = rem.slice(lDigits)
      if (!/^[0-9]*$/.test(snDigits)) throw new Error(`invalid nDigits: ${k}`);
      const nDigits = parseInt(snDigits, 10) + 10**(lDigits - 1);
      if (rem.length !== nDigits) throw new Error(`digit count mismatch: ${k}`);
      return BigInt(rem);
    }
    default:
      throw new Error(`invalid first character: ${k}`);
  }
};
```

</details>
</details>

Sample representations:

BigInt | serialization | explanation
-- | -- | --
0n | o | special case for positive–negative symmetry, correctly sorts after all negatives and before all positives
1n | p01 | p for positive, no following 9 so digit count is single-digit, 0 so digit count is 1, digits are "1"
10n | p110 | p for positive, no following 9 so digit count is single-digit, 1 so digit count is 2, digits are "10"
100n | p2100 | p for positive, no following 9 so digit count is single-digit, 2 so digit count is 3, digits are "100"
123456789n | p8123456789 | p for positive, no following 9 so digit count is single-digit, 8 so digit count is 9, digits are "123456789"
1234567890n | p9001234567890 | p for positive, one following 9 so digit count is two-digit, 00 so digit count is 10, digits are "1234567890"
12345678901n | p90112345678901 | p for positive, one following 9 so digit count is two-digit, 01 so digit count is 11, digits are "12345678901"
934857932847598725662n | p911934857932847598725662 | p for positive, one following 9 so digit count is two-digit, 11 so digit count is 21, digits are "934857932847598725662"
-1n | n98 | n for negative, no following 0 so digit count is single-digit, 9 so digit count is 1, digits are nines' complement of "8" which is "1"
-10n | n889 | n for negative, no following 0 so digit count is single-digit, 8 so digit count is 2, digits are nines' complement of "89" which is "10"
-100n | n7899 | n for negative, no following 0 so digit count is single-digit, 7 so digit count is 3, digits are nines' complement of "899" which is "100"
-123456789n | n1876543210 | n for positive, no following 0 so digit count is single-digit, 1 so digit count is 9, digits are nines' complement of "876543210" which is "123456789"
-1234567890n | n0998765432109 | n for negative, one following 0 so digit count is two-digit, 99 so digit count is 10, digits are nines' complement of "8765432109" which is "1234567890"
-12345678901n | n09887654321098 | n for negative, one following 0 so digit count is two-digit, 98 so digit count is 11, digits are nines' complement of "87654321098" which is "12345678901"
-934857932847598725662n | n088065142067152401274337 | n for negative, one following 0 so digit count is two-digit, 88 so digit count is 21, digits are nines' complement of "065142067152401274337" which is "934857932847598725662"


Control characters could be added for faster human comprehension, but are not necessary.

**Comment by erights** [comment]
Hi @gibson042 , https://github.com/Agoric/agoric-sdk/pull/4469#issuecomment-1030770373 is great!

I also appreciate the documentation examples there and the tests at https://jsfiddle.net/o6vw2nfd/ . Would be wonderful to see all this become part of the PR before merging.

I substituted in your code and added TODO placeholders for the rest. Since at this point, the contents of the PR will be almost all your contributions rather than mine, could you take it from here? I added you as co-assignee. Thanks!

**Comment by erights** [comment]
Because I'm the original author, github won't let me officially review. But consider me to be added unofficially to the reviewers.

**Comment by gibson042** [comment]
@erights To what extent does compactness matter? These could be made shorter with hexadecimal (and shorter still with base64, needing only four characters total for values up to 4095 and five for values up to 262143), but it would no longer be easy for a human to decode positive values.

**Comment by erights** [comment]
Would always be less that 2x difference?

At this stage, I'd say that it need not be shorter. Some human inspectability has some value, though is also imperfect and misleading. At this state, my inclination is to keep the decimal digits and even put back in the second colon that you mention. But not the first, since the numbers to the left of the colon no longer have any obvious relationship to anything.

Any further shrinkage probably should be driven by measurements telling us it would actually make a difference. So not yet.

File: packages/store/src/patterns/encodeKey.js
Comment (gibson042) [inline packages/store/src/patterns/encodeKey.js]: Is it reasonable to include this kind of assertion, and especially to include sequential assertions as more of the input is consumed? One alternative would be extracting all the parts up front with a single regular expression match like
```
const m = /^(?<prefix>p~*|n#*)(?<nDigits>[0-9]+):(?<n>[0-9]+)$/.exec(encodedKey);
```
But that's more of an all-or-nothing approach that is less likely to produce good messages when fed bad input.

File: packages/store/test/test-encodeKey.js
Comment (gibson042) [inline packages/store/test/test-encodeKey.js]: Does anyone have preferences on this nines' complement representation vs. ten's complement?

File: packages/store/test/test-encodeKey.js
Comment (gibson042) [inline packages/store/test/test-encodeKey.js]: Those tests sample _n_ of each _n_-digit range plus an adjacent neighborhood using Math.random(), is that acceptable here?

File: packages/store/src/patterns/encodeKey.js
Comment (gibson042) [inline packages/store/src/patterns/encodeKey.js:93]: It would also be possible to use a single type character for all bigint values rather than "n" for negative and "p" for nonnegative, e.g.

n | encoded
-- | --
-1n | i#8:8
0n | i~1:0
1n | i~1:1
5n | i~1:5
42n | i~2:42
1234567890n | i~~10:1234567890

**Review by gibson042 — COMMENTED** [review]
@erights I think this is ready for your review.

File: packages/store/src/patterns/encodeKey.js
Comment (erights) [inline packages/store/src/patterns/encodeKey.js:54]: Good choice moving the initial `f` that way.

File: packages/store/src/patterns/encodeKey.js
Comment (erights) [inline packages/store/src/patterns/encodeKey.js]: Yes, I think that would be better.

File: packages/store/src/patterns/encodeKey.js
Comment (erights) [reply packages/store/src/patterns/encodeKey.js:93]: What's the advantage? If none, I prefer `n` and `p`. But I have no strong preference. If your strong preference is `i`, go with it.

File: packages/store/src/patterns/encodeKey.js
Comment (erights) [reply packages/store/src/patterns/encodeKey.js]: More precise error messages are good. I like it.

File: packages/store/test/test-encodeKey.js
Comment (erights) [reply packages/store/test/test-encodeKey.js]: In my ignorance, I have no opinion. What are the pros and cons?

File: packages/store/test/test-encodeKey.js
Comment (erights) [reply packages/store/test/test-encodeKey.js]: Not using `Math.random()`.

@dckc @warner @mhofman what do we do for alleged randomness in tests? What should @gibson042 do here?

@gibson042 if this question isn't trivially answered, proceed with this PR without them. But it would be good for those tests to follow in a later PR.

**Review by erights — COMMENTED** [review]
LGTM, but I can't approve. However, you already have an approval from @FUDCo . I think it's ok to proceed on that.

Note to @FUDCo , we're using your approval even though substantial changes since then. OTOH, I would approve if I could.

File: packages/store/test/test-encodeKey.js
Comment (gibson042) [reply packages/store/test/test-encodeKey.js]: Basically just subjective ease vs. difficulty of decoding negative numbers as a human. This encoding uses only 90% of all n-digit sequences (because the remaining 10% start with zero and are therefore actually sequences numbers of _fewer_ than n digits), which can therefore be offset according to taste.

n | nines' complement | ten's complement
-- | -- | --
-1n | n8:8 | n9:9
-9n | n8:0 | n9:1
-10n | n7:89 | n8:90
-11n | n7:88 | n8:89
-99n | n7:00 | n8:01
-100n | n6:899 | n7:900
-900n | n6:099 | n7:100
-999n | n6:000 | n7:001
-1234567890n | n#89:8765432109 | n#90:8765432110

File: packages/store/src/patterns/encodeKey.js
Comment (gibson042) [reply packages/store/src/patterns/encodeKey.js:93]: There's no significant difference, it just seemed worth raising in case of perceived value from e.g. economy of type prefixes and/or one-to-one mappings. Assuming not, we'll stick with "n" and "p".

File: packages/store/test/test-encodeKey.js
Comment (erights) [reply packages/store/test/test-encodeKey.js]: 10s complement looks easier to me to visually decode. If you agree, let's do that.

File: packages/store/src/patterns/encodeKey.js
Comment (erights) [reply packages/store/src/patterns/encodeKey.js:93]: good

File: packages/store/test/test-encodeKey.js
Comment (gibson042) [reply packages/store/test/test-encodeKey.js]: Done.

File: packages/store/test/test-encodeKey.js
Comment (dckc) [reply packages/store/test/test-encodeKey.js]: > @dckc @warner @mhofman what do we do for alleged randomness in tests? What should @gibson042 do here?

That looks like property testing. We have 2 precedents for using [fast-check](https://www.npmjs.com/package/fast-check) to generate arbitrary / random inputs.

https://github.com/Agoric/agoric-sdk/blob/79c4276da8c856674bd425c54715adec92648c48/packages/ERTP/test/unitTests/test-amountProperties.js

https://github.com/Agoric/agoric-sdk/blob/a5de8dd298f2d7226cd4e8c9a3de69fe493c630b/packages/run-protocol/test/amm/vpool-xyk-amm/test-amm-calc.js

Comment (gibson042) [reply packages/store/test/test-encodeKey.js]: Thanks, I've added tests after much wrestling with fast-check (and still can't get encodeKey equivalents working, but I'll open a separate PR for that).

File: packages/store/test/test-encodeKey.js
Comment (gibson042) [inline packages/store/test/test-encodeKey.js]: Reproducing logic like this is super kludgy, but ava assertions returning a success/failure boolean was only added in v4 (cf. https://github.com/avajs/ava/releases/tag/v4.0.0#:~:text=Assertions%20as%20type%20guards ) and the upgrade appears to be non-trivial.

File: packages/store/test/test-encodeKey.js
Comment (dckc) [inline packages/store/test/test-encodeKey.js]: is this equivalent and more concise?

```suggestion
      const ea = encodeKey(a);
      const eb = encodeKey(b);
      t.is(a === b, ea === eb);
      t.is(a < b, ea < eb);
      t.is(a > b, ea > eb);
```

File: packages/store/test/test-encodeKey.js
Comment (dckc) [inline packages/store/test/test-encodeKey.js:132]: Is it important to return something from the 2nd arg of `fc.property`? Doesn't this suffice?

```suggestion
      t.is(rt, n);
```

Comment (gibson042) [reply packages/store/test/test-encodeKey.js:132]: The examples at https://github.com/dubzzz/fast-check/#getting-started made it look like returning true(thy) was important, but https://dubzzz.github.io/fast-check/index.html#property suggests that only exactly `false` or an exception are treated as failures. I'll update accordingly.

Comment (gibson042) [reply packages/store/test/test-encodeKey.js:132]: Hmm, the problem with skipping `return` is that doing so bypasses the fast-check behavior for presenting a reduced counterexample to the property being asserted—rather than something useful like
```
  encodeKey › BigInt encoding comparison corresponds with numeric comparison

  Rejected promise returned by test. Reason:

  Error {
    message: `Property failed after 6 tests␊
    { seed: 1977596688, path: "5:71:124", endOnFailure: true }␊
    Counterexample: [35n,-23836150654801173763396695620220553254412511975658213618617502216148950167540n]␊
    Shrunk 2 time(s)␊
    Got error: Property failed by returning false␊
    ␊
    Hint: Enable verbose mode in order to have the list of all failing values encountered during the run`,
  }
```
we just get
```
  encodeKey › BigInt encoding comparison corresponds with numeric comparison

  Difference:

  - true
  + false
```

I think I'd like to keep `assertionPassed` until it is made superfluous by upgrade to ava v4+ (in which assertions return a "passed" boolean, allowing replacement of `return assertionPassed(t.<method>(…), () => …)` with `return t.<method>(…)`).

File: packages/store/test/test-encodeKey.js
Comment (gibson042) [reply packages/store/test/test-encodeKey.js]: Not quite, but I like the general approach. Adopted.

File: packages/store/test/test-encodeKey.js
Comment (dckc) [reply packages/store/test/test-encodeKey.js:132]: oh wow! Thanks! I was kinda disappointed that fast-check didn't seem to do the usual shrinking thing. Good to know I was just holding it wrong.

**Pull request merged**
